kernel/rros/clock.rs

```diff
@@ -336,41 +331,47 @@ pub fn adjust_timer(
 }
 
 pub fn rros_adjust_timers(clock: &mut RrosClock, delta: KtimeT) {
-    // Adjust all timers in the List in each CPU tmb of the current clock.
-    // raw_spin_lock_irqsave(&tmb->lock, flags);
-    let cpu = 0;
-    // for_each_online_cpu(cpu) {
-    let rq = rros_cpu_rq(cpu);
-    let tmb = rros_percpu_timers(clock, cpu);
-    let tq = unsafe { &mut (*tmb).q };
-
-    for i in 1..=tq.len() {
-        let timer = tq.get_by_index(i).unwrap().value.clone();
-        let get_clock = timer.lock().get_clock();
-        if get_clock == clock as *mut RrosClock {
-            rros_dequeue_timer(timer.clone(), tq);
-            adjust_timer(clock, timer.clone(), tq, delta);
+    for cpu in online_cpus() {
+        let rq: *mut rros_rq = rros_cpu_rq(cpu as i32);
+        let tmb = rros_percpu_timers(clock, cpu as i32);
+        let tq = unsafe { &mut (*tmb).q };
+        
+        let flags:u64 = unsafe { (*tmb).lock.irq_lock_noguard() };
+
+        for i in 1..=tq.len() {
+            let timer = tq.get_by_index(i).unwrap().value.clone();
```

Comment:
I am a little confused. I noticed you are using the index and removing the element at the same time, is it safe? Or should we use an iterator instead ( Maybe the `retain` is a good choice ). The list still needs to be refactored.

Reply:
Yes, here is a problem. I didn't discover it. Thanks!